[Yuto Watanabe] iP

yuto1115 opened the pull request:
# Odin

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Odin frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/yuto1115/ip/releases).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x]  Managing to-dos
- [x]  Managing deadlines
- [x]  Managing events
- [ ]  Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        new Odin("./src/data/records.txt").run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

yuto1115 committed:
Rename, Greet, Exit

yuto1115 committed:
Echo

yuto1115 committed:
add task add/list feature

yuto1115 committed:
add Task class

yuto1115 committed:
change method names

yuto1115 committed:
fix Task class

yuto1115 committed:
add mark/unmark feature

yuto1115 committed:
change Task to abstract class

yuto1115 committed:
fix Task

yuto1115 committed:
add todo/deadline/event

yuto1115 committed:
write text-ui-test

yuto1115 committed:
fix a bug

yuto1115 committed:
implement error handling using Exception

yuto1115 committed:
add delete feature

yuto1115 committed:
Refactor: Delete indent before "case"

yuto1115 committed:
Add javadoc for methods in Odin.java

yuto1115 committed:
Change method name to align with naming convention

yuto1115 committed:
Add record file to .gitignore

yuto1115 committed:
Add feature to save and recover task list

yuto1115 committed:
Update EXPECTED.TXT

yuto1115 committed:
Merge branch 'branch-Level-7'

yuto1115 committed:
Change exception handling scheme

yuto1115 committed:
Create DateAndOptionalTime class and use it to specify date&time

yuto1115 committed:
Update text-ui-test

yuto1115 committed:
Merge branch 'branch-Level-8'

yuto1115 committed:
Refactor code by organising them into separated classes and packages

yuto1115 committed:
Merge branch 'add-gradle-support'

yuto1115 committed:
Set entry point in build.gradle

yuto1115 committed:
Refactor Task.java: change List.of to Arrays.asList

yuto1115 committed:
Support JavaFX

yuto1115 committed:
Remove unnecessary quote

yuto1115 committed:
Fix AddEventCommand.getCommandName

yuto1115 committed:
Write unit test for DateAndOptionalTime

yuto1115 committed:
Write unit test for Parser

yuto1115 committed:
Update build.gradle: change archiveBaseName of shadowJar

yuto1115 committed:
Refactor code using IntelliJ

yuto1115 committed:
Add JavaDoc for every public class/method

yuto1115 committed:
Change the access modifier of 'done' in Task class

yuto1115 committed:
Add find command

yuto1115 committed:
Merge branch 'branch-A-JavaDoc'

yuto1115 committed:
Merge branch 'branch-A-CodingStandard'

yuto1115 committed:
Resolve merge conflict

yuto1115 committed:
Merge branch 'branch-Level-9'

yuto1115 committed:
Update .gitignore

yuto1115 committed:
Set up Checkstyle

yuto1115 committed:
Refactor codes by according to Checkstyle

yuto1115 committed:
Add a suppression in suppressions.xml: suppress MethodName warning for test files

yuto1115 committed:
Merge branch 'branch-A-CheckStyle'

vrisdng commented on src/main/java/odin/parser/DateAndOptionalTime.java:
```diff
@@ -0,0 +1,63 @@
+        try {
+            this.date = LocalDate.parse(tokens.get(0));
+        } catch (DateTimeParseException e) {
+            throw new WrongFormatException("Date cannot be parsed as yyyy-MM-dd.");
```
Do you think it's better to make this more generalized? I think that not only specific format is unable to be parsed, but other potential format might also have errors. 

My suggestion is telling the user that date cannot be parsed and give the user the correct format they need to follow.

vrisdng commented on src/main/java/odin/parser/command/Command.java:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Interface that represents an user command.
+ */
+public interface Command {
```
Nice OOP for Command here!

vrisdng commented on src/main/java/odin/task/Task.java:
```diff
@@ -0,0 +1,119 @@
+    static Task restoreFromTaskRecord(ArrayList<String> taskRecord) throws WrongFormatException {
+        int len = taskRecord.size();
+        if (len < 3) {
+            throw new WrongFormatException("Number of tokens in a Odin.task record must be at least three.");
```
The use of word like "tokens" might be a bit confusing for common users. Do you want to change it to a more direct and clearer way? e.g: Giving them the correct format, like my previous comment.

vrisdng reviewed (commented):
Overall, your code adheres to OOP nicely and well-organized!

hyizhak commented on src/main/java/odin/parser/command/FindCommand.java:
```diff
@@ -0,0 +1,63 @@
+ * Class for the command to find tasks by searching for a keyword.
+ */
+public class FindCommand implements Command {
+    private TaskList taskList;
+    private String pattern;
+    private ArrayList<Integer> matchIndexes;
```
These instance variables (taskList, pattern, matchIndexes) seem to store state unnecessarily. Is FindCommand meant to be a stateless command? And I think this applies to most of the command related classes.

hyizhak commented on src/main/java/odin/parser/Parser.java:
```diff
@@ -0,0 +1,68 @@
+        String commandName = tokens.get(0).toLowerCase();
+        ArrayList<String> arguments = new ArrayList<>(tokens.subList(1, tokens.size()));
+
+        Command[] commands = {
```
Do you think it's a bit inefficient that every time parseAndHandle is called, a new instance of each command is created?

hyizhak reviewed (commented):
Overall looks good to me. Just a few nits to be fixed.

yuto1115 committed:
Fix exception messages

yuto1115 committed:
Modify the entry point

yuto1115 committed:
Create a launcher class

yuto1115 committed:
Remove unnecessary print of separator

yuto1115 committed:
Add images for GUI

yuto1115 committed:
Switch from CUI to GUI

yuto1115 committed:
Follow naming convention for booleans

yuto1115 committed:
Fix header comments

yuto1115 committed:
Refactor and refine GUI

yuto1115 committed:
Add final modifier

yuto1115 committed:
Merge branch 'branch-Level-10'

yuto1115 committed:
Add assertions to check validity of index

There are multiple command classes with a method that takes in indexes as arguments, without checking their validity.

Even though these indexes are guaranteed to be valid if everything is implemented correctly, it is better to use assertions in order to ensure and demonstrate the validity.

Let's put assertions, in every function that takes in indexes as arguments, to check their validity.

yuto1115 committed:
Refactor: decompose the `restoreFromTaskRecord` method

The `restoreFromTaskRecord` first checks task type, does switch-case by the type, and describes different processes needed to parse the task record for each task type.

This implementation impairs readability due to its deep nesting, and also breaks the SLAP principle.

Let's make independent methods for processing each task type, and let them be called from the switch-case of `restoreFromTaskRecord`.

yuto1115 committed:
Merge pull request #3 from yuto1115/branch-A-Assertions

Add assertions to check validity of index

yuto1115 committed:
Merge branch 'master' into branch-A-CodeQuality

yuto1115 committed:
Merge pull request #2 from yuto1115/branch-A-CodeQuality

Refactor: decompose the `restoreFromTaskRecord` method

yuto1115 committed:
Fix the javafx version in fxml file

yuto1115 committed:
Change WrongFormatException message

yuto1115 committed:
Disable [un]mark command if it's already [un]marked

yuto1115 committed:
Add undo command

yuto1115 committed:
Merge branch 'branch-C-Undo'

yuto1115 committed:
Add margin between icon and text

yuto1115 committed:
Set title on the GUI window

yuto1115 committed:
Remove unnecessary output of separator

yuto1115 committed:
Add UI.png

yuto1115 committed:
Rename UI.png to Ui.png

yuto1115 committed:
Write user guide

yuto1115 committed:
Modify user guide
